File: 00190. Reverse Bits/solution.cpp
/**
 * Problem: 190. Reverse Bits
 *
 * Difficulty: Easy
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution {
public:
    int reverseBits(int n) {
        // Initialize result to 0
        int reversed = 0;

        // Process all 32 bits of the input number
        for (int i = 0; i < 32; i++) {
            // Shift result left by 1 to make room for next bit
            // Extract i-th bit from input using right shift and mask
            // Combine using OR operation
            reversed = (reversed << 1) | ((n >> i) & 1);
        }

        // Return number with reversed bits
        return reversed;
    }
};

File: 00190. Reverse Bits/solution.js
/**
 * Problem: 190. Reverse Bits
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Reverses the bits of a 32-bit unsigned integer
 *
 * @param {number} n - Input 32-bit unsigned integer
 *
 * @return {number} Number with reversed bits
 */
const reverseBits = n => {
    // Array to store bits in reverse order
    let reversedBits = []

    // Extract all 32 bits from input number
    for (let i = 0; i < 32; i++) {
        // Get least significant bit
        let remainder = n % 2

        // Shift number right by 1 bit
        n = Math.floor(n / 2)

        // Store bit in array
        reversedBits.push(remainder)
    }

    // Reconstruct number from reversed bits
    let result = 0

    for (let bit of reversedBits)
        // Shift left and add current bit
        result = result * 2 + bit

    // Return number with reversed bits
    return result
}

const fs = require("fs")

process.on("exit", () => { fs.writeFileSync("display_runtime.txt", "0") })
